README.md
<p align="center">

![](laravel-postal-code-validation.png)

</p>

# LaravelPostcodes

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-github-action]][link-github-action]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

A service wrapper around [postcodes.io](postcodes.io) with validation rule and macro

## Install

Via Composer

```bash
$ composer require juststeveking/laravel-postcodes
```

After installation, merge configuration for services uusing:

```bash
$ php artisan vendor:publish --provider="JustSteveKing\LaravelPostcodes\PostcodesServiceProvider"
```

If, for some reason, this doesn't work please use the following steps:

- Add the following into the `config/services.php` configuration file:

```php
<?php

'postcodes' => [
    'url' => env('POSTCODES_URL', 'https://api.postcodes.io/')
],
```

- Add `POSTCODES_URL` to your `.env` file and add `https://api.postcodes.io/` as the value.


## Basic Usage

You can use the validation rule:

``` php
<?php

$this->validate($request, [
    'postcode' => [
        'required',
        'string',
        new Postcode()
    ]
]);
```

Or you can use the validation Macro:

```php
$this->validate($request, [
    'postcode' => [
        'required',
        'string',
        Rule::postcode()
    ]
]);
```

If you want to interact with the service itself

```php
<?php 

use JustSteveKing\LaravelPostcodes\Service\PostcodeService;

class SomeController extends Controller
{
    protected $postcodes;

    public function __construct(PostcodeService $service)
    {
        $this->postcodes = $service;
    }

    public function store(Request $request)
    {
        // validation using example above
        $location = $this->postcodes->getPostcode($request->postcode);
    }
}
```

### Validate

```php
<?php

$service = resolve(PostcodeService::class);

$service->validate('AB10 1AB');
```

### Validate Postcode

```php
<?php

$service = resolve(PostcodeService::class);

$service->validate('AB10 1AB');
```

### Get Postcode information

```php
<?php

$service = resolve(PostcodeService::class);

$service->getPostcode('AB10 1AB');
```


### Bulk Lookup Postcodes

```php
<?php

$service = resolve(PostcodeService::class);

$service->getPostcodes([
    'AB10 1AB',
    'AB10 1AF',
    'AB10 1AG',
]);
```

### Get nearest postcodes for a given longitude & latitude

```php
<?php

$service = resolve(PostcodeService::class);

$service->nearestPostcodesForGivenLngAndLat(
    0.629806,
    51.792326
);
```

### Nearest postcodes for postcode

```php
<?php

$service = resolve(PostcodeService::class);

$service->nearest('AB10 1AB');
```

### Autocomplete a postcode partial

```php
<?php

$service = resolve(PostcodeService::class);

$service->autocomplete('AB10');
```

### Query for postcode

```php
<?php

$service = resolve(PostcodeService::class);

$service->query('AB10 1AB');
```

### Lookup terminated postcode

```php
<?php

$service = resolve(PostcodeService::class);

$service->getTerminatedPostcode('AB1 0AA');
```

### Lookup Outward Code

```php
<?php

$service = resolve(PostcodeService::class);

$service->getOutwardCode('N11');
```

### Nearest outward code for outward code

```php
<?php

$service = resolve(PostcodeService::class);

$limit = 80; // Limit needs to be less than 100
$radius = 15000; // Radius needs to be less than 25000
$service->getNearestOutwardCode('N11', $limit, $radius);
```

### Get nearest outward codes for a given longitude & latitude

```php
<?php

$service = resolve(PostcodeService::class);

$service->nearestOutwardCodesForGivenLngAndLat(
    0.629806,
    51.792326
);
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email juststevemcd@gmail.com instead of using the issue tracker.

## Credits

- [Steve McDougall][link-author]
- [All Contributors][link-contributors]
- [Laravel News for the artiwork](https://www.laravel-news.com)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/juststeveking/laravel-postcodes.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-github-action]: https://github.com/JustSteveKing/LaravelPostcodes/workflows/build-tests/badge.svg?branch=master
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/JustSteveKing/LaravelPostcodes.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/JustSteveKing/LaravelPostcodes.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/juststeveking/laravel-postcodes.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/juststeveking/laravel-postcodes
[link-github-action]: https://github.com/JustSteveKing/LaravelPostcodes/actions
[link-scrutinizer]: https://scrutinizer-ci.com/g/JustSteveKing/LaravelPostcodes/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/JustSteveKing/LaravelPostcodes
[link-downloads]: https://packagist.org/packages/juststeveking/laravel-postcodes
[link-author]: https://github.com/JustSteveKing
[link-contributors]: ../../contributors

CHANGELOG.md
# Changelog

All notable changes to `LaravelPostcodes` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## v1.2.0 - 2020-03-03

### Added
- A Pre validate option to validate, allowing regex matching to save service calls
- A regex based outcode validation, this was a PR by PHPAdam that needed some work
- Tests for both of the above
- Laravel 7.* support

## v1.1.0 - 2019-10-25

### Added
- Make tests more stringent
- Unit tests make API calls 
- Update Postcode Service
- Bulk lookup postcodes
- Get nearest postcodes for a given longitude & latitude
- Nearest postcodes for postcode
- Autocomplete a postcode partial
- Query for postcode
- Lookup terminated postcode
- Lookup Outward Code
- Nearest outward code for outward code
- Get nearest outward codes for a given longitude & latitude
- Convert all returns to return Collection instead off Array where appropriate

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/TestCase.php
<?php

declare(strict_types=1);

namespace JustSteveKing\LaravelPostcodes;

class TestCase extends \Orchestra\Testbench\TestCase
{
    protected function setUp(): void
    {
        parent::setUp();

        try {
            $env = parse_ini_file(__DIR__ . '/../.env');
            if (isset($env['POSTCODES_URL'])) {
                $this->app['config']->set('services.postcodes.url', $env['POSTCODES_URL']);
            }
        } catch (\Exception $e) {
            //
        }

        if (!defined('LARAVEL_START')) {
            define('LARAVEL_START', microtime(true));
        }
    }

    protected function getPackageProviders($app)
    {
        return [
            \JustSteveKing\LaravelPostcodes\PostcodesServiceProvider::class
        ];
    }
}


tests/Unit/PostcodeRuleTest.php
<?php

declare(strict_types=1);

namespace JustSteveKing\LaravelPostcodes\Unit;

use GuzzleHttp\Client;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use Illuminate\Validation\Rule;
use JustSteveKing\LaravelPostcodes\Service\PostcodeService;
use JustSteveKing\LaravelPostcodes\TestCase;
use JustSteveKing\LaravelPostcodes\Rules\Postcode;

class PostcodeRuleTest extends TestCase
{
    /**
     * @var Postcode
     */
    protected $rule;

    public function testPostcodeMacro()
    {
        $this->setupRule(200);

        $this->assertInstanceOf(Postcode::class, Rule::postcode());
    }

    public function testValidPostcodePasses()
    {
        $this->setupRule(200, json_encode(['result' => true]));

        $postcode = 'CF10 4UW';
        $this->assertTrue($this->rule->passes('postcode', $postcode));
    }

    public function testInvalidPostcodeFails()
    {
        $this->setupRule(200, json_encode(['result' => false]));

        $postcode = 'testing';
        $this->assertFalse($this->rule->passes('postcode', $postcode));
    }

    public function testValidationMessageIsCorrect()
    {
        $this->setupRule(200);

        $string = 'The submitted postcode is not a valid UK postcode';

        $this->assertEquals($this->rule->message(), $string);
    }

    private function setupRule(int $status, string $body = null): void
    {
        $mock = new MockHandler([new Response($status, [], $body)]);
        $handler = HandlerStack::create($mock);
        $client = new Client(['handler' => $handler]);

        $this->rule = new Postcode(new PostcodeService($client));
    }
}


tests/Unit/PostcodeServiceTest.php
<?php

declare(strict_types=1);

namespace JustSteveKing\LaravelPostcodes\Unit;

use GuzzleHttp\Client;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use JustSteveKing\LaravelPostcodes\TestCase;
use JustSteveKing\LaravelPostcodes\Service\PostcodeService;

class PostcodeServiceTest extends TestCase
{
    /** @var MockHandler */
    protected $handler;

    protected $postcode = 'N11 1QZ';
    protected $terminatedPostcode = 'AB1 0AA';

    public function testServiceIsCorrectType()
    {
        $this->assertInstanceOf(PostcodeService::class, $this->service(200));
    }

    public function testServiceCanValidatePostcode()
    {
        $serviceFail = $this->service(200, json_encode(['result' => false]));
        $this->assertEquals(false, $serviceFail->validate('test'));
        $this->assertRequest('GET', 'https://api.postcodes.io/postcodes/test/validate');

        $serviceSuccess = $this->service(200, json_encode(['result' => true]));
        $this->assertTrue($serviceSuccess->validate($this->postcode));
        $this->assertRequest('GET', 'https://api.postcodes.io/postcodes/N11 1QZ/validate');
    }

    public function testServiceCanValidatePostcodeUsingPrevalidateOption()
    {
        $service = $this->service(200, json_encode(['result' => ['postcode' => $this->postcode]]));
        $result = $service->validate($this->postcode, true);
        $this->assertTrue($result);
        
        $fails = $service->validate('fake-postcode', true);
        $this->assertEquals(false, $fails);
    }

    public function testServiceCanValidateAnOutcode()
    {
        $service = $this->service(200, json_encode(['result' => ['postcode' => $this->postcode]]));
        $result = $service->validateOutcode($this->postcode);
        $this->assertTrue($result);

        $result = $service->validateOutcode("123123123123");
        $this->assertEquals(false, $result);
    }

    public function testServiceCanGetPostcode()
    {
        $service = $this->service(200, json_encode(['result' => ['postcode' => $this->postcode]]));
        $result = $service->getPostcode($this->postcode);

        $this->assertEquals($result->postcode, $this->postcode);
        $this->assertRequest('GET', 'https://api.postcodes.io/postcodes/N11 1QZ');
    }

    public function testServiceCanGetOutwardCode()
    {
        $service = $this->service(200, json_encode(['result' => ['outcode' => substr($this->postcode, 0, 3)]]));
        $result = $service->getOutwardCode(substr($this->postcode, 0, 3));

        $this->assertEquals($result->outcode, substr($this->postcode, 0, 3));
    }

    public function testServiceCanGetRandomPostcode()
    {
        $service = $this->service(200, json_encode(['result' => ['postcode' => $this->postcode]]));
        $result = $service->getRandomPostcode();

        $this->assertNotNull($result->postcode);
        $this->assertRequest('GET', 'https://api.postcodes.io/random/postcodes');
    }

    public function testServiceCanQueryPostcode()
    {
        $serviceFound = $this->service(200, json_encode(['result' => [['postcode' => $this->postcode]]]));
        $resultFound = $serviceFound->query($this->postcode);

        $this->assertInstanceOf(\Illuminate\Support\Collection::class, $resultFound);
        $this->assertCount(1, $resultFound);
        $this->assertRequest('GET', 'https://api.postcodes.io/postcodes?q=N11 1QZ');

        $serviceNull = $this->service(200, json_encode(['result' => null]));
        $resultNull = $serviceNull->query('AA1 1AA');

        $this->assertTrue($resultNull->isEmpty());
        $this->assertRequest('GET', 'https://api.postcodes.io/postcodes?q=AA1 1AA');
    }

    public function testServiceCanGetTerminatedPostcode()
    {
        $service = $this->service(200, json_encode([
            'result' => [
                'postcode'         => $this->terminatedPostcode,
                'year_terminated'  => 1996,
                'month_terminated' => 6,
                'longitude'        => -2.242851,
                'latitude'         => 57.101474,
            ],
        ]));
        $result = $service->getTerminatedPostcode($this->terminatedPostcode);

        $this->assertNotNull($result->postcode);
        $this->assertRequest('GET', 'https://api.postcodes.io/terminated_postcodes/AB1 0AA');
    }

    public function testServiceCanAutocompletePostcode(): void
    {
        $data = [
            'status' => 200,
            'result' => [
                "AB10 1AB",
                "AB10 1AF",
                "AB10 1AG",
                "AB10 1AH",
                "AB10 1AL",
                "AB10 1AN",
                "AB10 1AP",
                "AB10 1AQ",
                "AB10 1AR",
                "AB10 1AS",
            ],
        ];
        $service = $this->service(200, json_encode($data));
        $partialPostcode = 'some-postcode-with-autocomplete-results';

        $actual = $service->autocomplete($partialPostcode);

        $this->assertSame($data['result'], $actual->toArray());
        $this->assertRequest(
            'GET',
            'https://api.postcodes.io/postcodes/some-postcode-with-autocomplete-results/autocomplete'
        );
    }

    public function testServiceCantAutocompletePostcode(): void
    {
        $data = [
            'status' => 200,
            'result' => null,
        ];
        $service = $this->service(200, json_encode($data));
        $partialPostcode = 'some-postcode-without-autocomplete-results';

        $actual = $service->autocomplete($partialPostcode);

        $this->assertTrue($actual->isEmpty());
        $this->assertRequest(
            'GET',
            'https://api.postcodes.io/postcodes/some-postcode-without-autocomplete-results/autocomplete'
        );
    }

    public function testServiceCanGetNearestPostcodes()
    {
        $serviceFound = $this->service(
            200,
            json_encode(['result' => [['postcode' => $this->postcode]]])
        );

        $resultFound = $serviceFound->nearest($this->postcode);

        $this->assertInstanceOf(\Illuminate\Support\Collection::class, $resultFound);
        $this->assertCount(1, $resultFound);
    }

    public function testServiceCanGetPostcodes()
    {
        $postcodes = ["PR3 0SG", "M45 6GN", "EX165BL"];
        $service = $this->service(200, json_encode([
            "status" => 200,
            "result" => [
                ["query" => "PR3 0SG", "result" => ["postcode" => "PR3 0SG",],],
                ["query" => "M45 6GN", "result" => ["postcode" => "M45 6GN",],],
                ["query" => "EX165BL", "result" => ["postcode" => "EX16 5BL"]]
            ]
        ]));
        $result = $service->getPostcodes($postcodes, ['postcode']);

        $this->assertInstanceOf(\Illuminate\Support\Collection::class, $result);
        $this->assertEquals($result->count(), 3);
        $this->assertEquals($result->first()->postcode, $postcodes[0]);
    }

    public function testServiceCanGetNearestOutwardCodesForGivenLongitudeAndLatitude()
    {
        $json = file_get_contents(
            __DIR__
            . '/../Fixtures/GetNearestOutwardCodesForGivenLongitudeAndLatitude.json'
        );
        $serviceFound = $this->service(200, $json);

        $expected = json_encode(json_decode($json)->result);
        $actual = json_encode($serviceFound->nearestOutwardCodesForGivenLngAndLat(0.629834723775309, 51.7923246977375));

        $this->assertSame($expected, $actual);
    }

    public function testServiceCanHandleEmptyResponseForNearestOutwardCodesForGivenLongitudeAndLatitude()
    {
        $serviceFound = $this->service(
            200,
            json_encode(['result' => null])
        );

        $actual = $serviceFound->nearestOutwardCodesForGivenLngAndLat(0, 0);

        $this->assertTrue($actual->isEmpty());
    }

    public function testServiceCanGetNearestOutwardCode()
    {
        $service = $this->service(200, json_encode(['result' => [['outcode' => substr($this->postcode, 0, 3)]]]));
        $result = $service->getNearestOutwardCode(substr($this->postcode, 0, 3));

        $this->assertInstanceOf(\Illuminate\Support\Collection::class, $result);
        $this->assertEquals($result->first()->outcode, substr($this->postcode, 0, 3));
    }

    public function testServiceCanGetNearestPostcodesForGivenLngAndLat()
    {
        $json = file_get_contents(
            __DIR__
            . '/../Fixtures/GetNearestPostcodesForGivenLongitudeAndLatitude.json'
        );

        $serviceFound = $this->service(200, $json);

        $expected = json_encode(json_decode($json)->result);

        $actual = json_encode(
            $serviceFound->nearestPostcodesForGivenLngAndLat(0.629834723775309, 51.7923246977375)
                ->toArray()
        );

        $this->assertSame($expected, $actual);
    }

    public function testServiceCanHandleEmptyResponseForCanGetNearestPostcodesForGivenLngAndLat()
    {
        $serviceFound = $this->service(
            200,
            json_encode(['result' => null])
        );

        $actual = $serviceFound->nearestPostcodesForGivenLngAndLat(0, 0);

        $this->assertTrue($actual->isEmpty());
    }

    private function service(int $status, string $body = null): PostcodeService
    {
        $this->handler = new MockHandler([new Response($status, [], $body)]);
        $stack = HandlerStack::create($this->handler);
        $client = new Client(['handler' => $stack]);

        return new PostcodeService($client);
    }

    private function assertRequest(string $method, string $uri): void
    {
        $request = $this->handler->getLastRequest();

        $this->assertSame($method, $request->getMethod());
        $this->assertSame($uri, urldecode((string)$request->getUri()));
    }
}


src/Rules/Postcode.php
<?php

declare(strict_types=1);

namespace JustSteveKing\LaravelPostcodes\Rules;

use Illuminate\Contracts\Validation\Rule;
use JustSteveKing\LaravelPostcodes\Service\PostcodeService;

class Postcode implements Rule
{
    protected $service;

    /**
     * Create a new rule instance.
     *
     * @return void
     */
    public function __construct(PostcodeService $service)
    {
        $this->service = $service;
    }

    /**
     * Determine if the validation rule passes.
     *
     * @param  string  $attribute
     * @param  mixed  $value
     * @return bool
     */
    public function passes($attribute, $value)
    {
        return $this->service->validate($value);
    }

    /**
     * Get the validation error message.
     *
     * @return string
     */
    public function message()
    {
        return 'The submitted postcode is not a valid UK postcode';
    }
}


src/PostcodesServiceProvider.php
<?php

declare(strict_types=1);

namespace JustSteveKing\LaravelPostcodes;

use GuzzleHttp\Client;
use Illuminate\Support\ServiceProvider;
use JustSteveKing\LaravelPostcodes\Service\PostcodeService;

class PostcodesServiceProvider extends ServiceProvider
{
    public function register()
    {
        $this->mergeConfigFrom(
            __DIR__ . '/../config/services.php',
            'services'
        );

        $this->mergeConfigFrom(
            __DIR__ . '/../config/postcodes.php',
            'postcodes'
        );

        $this->app->bind(PostcodeService::class, function () {
            return new PostcodeService(
                new Client()
            );
        });

        \Illuminate\Validation\Rule::macro('postcode', function () {
            return new \JustSteveKing\LaravelPostcodes\Rules\Postcode(resolve(PostcodeService::class));
        });
    }

    public function boot()
    {
        //
    }
}


src/Service/BulkReverseGeocoding/Geolocation.php
<?php

declare(strict_types=1);

namespace JustSteveKing\LaravelPostcodes\Service\BulkReverseGeocoding;

class Geolocation
{
    /** @var float */
    private $lng;
    /** @var float */
    private $lat;
    /** @var int|null */
    private $radius;
    /** @var int|null */
    private $limit;

    public function __construct(
        float $lng,
        float $lat,
        int $radius = null,
        int $limit = null
    ) {
        $this->lng = $lng;
        $this->lat = $lat;
        $this->radius = $radius;
        $this->limit = $limit;
    }

    public function getLng(): float
    {
        return $this->lng;
    }

    public function getLat(): float
    {
        return $this->lat;
    }

    public function getRadius(): ?int
    {
        return $this->radius;
    }

    public function getLimit(): ?int
    {
        return $this->limit;
    }

    public function toArray(): array
    {
        return [
            'longitude' => $this->getLng(),
            'latitude' => $this->getLat(),
        ] + array_filter([
            'radius' => $this->getRadius(),
            'limit' => $this->getLimit(),
        ]);
    }
}


src/Service/PostcodeService.php
<?php

declare(strict_types=1);

namespace JustSteveKing\LaravelPostcodes\Service;

use GuzzleHttp\Client;
use Illuminate\Support\Collection;
use function GuzzleHttp\Psr7\build_query;

class PostcodeService
{
    /**
     * @var string
     */
    protected $url;

    /**
     * @var Client
     */
    protected $http;

    /**
     * Postcode Service constructor.
     *
     * @param Client $client
     *
     * @return void
     */
    public function __construct(Client $client)
    {
        $this->url = config('services.postcodes.url');

        $this->http = $client;
    }

    /**
     * Validate a postcode against the API
     *
     * @param string $postcode
     * @param bool   $preValidate
     *
     * @return bool
     */
    public function validate(string $postcode, bool $preValidate = false): bool
    {
        if ($preValidate) {
            return !! preg_match(config('postcodes.regex.postcode'), $postcode);
        }

        return $this->getResponse("postcodes/$postcode/validate");
    }

    /**
     * Validate an Outcode using RegEx - not supported by the API
     * 
     * @param  string  $postcode
     * 
     * @return bool
     */
    public function validateOutcode(string $postcode): bool
    {
        return !! preg_match(config('postcodes.regex.outcode'), $postcode);
    }

    /**
     * Get the address details from a postcode
     *
     * @param string $postcode
     *
     * @return object
     */
    public function getPostcode(string $postcode): object
    {
        return $this->getResponse("postcodes/$postcode");
    }

    /**
     * Get the address details from a multiple postcodes at once
     *
     * @param array $postcodes
     *
     * @param array $filter - optional array of fields to return
     * @return Collection
     *
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    public function getPostcodes(array $postcodes, array $filter = []): Collection
    {
        if (!empty($filter)) {
            $filter = build_query(['filter' => implode(',', $filter)]);
        }

        return collect($this->getResponse(
            'postcodes?' . $filter,
            'POST',
            ['postcodes' => array_values($postcodes)]
        ))->map(function ($item) {
            return $item->result;
        });
    }

    /**
     * Get information based on outward code including geo data
     *
     * @param string $outwardcode
     *
     * @return object
     */
    public function getOutwardCode(string $outwardcode): object
    {
        return $this->getResponse("outcodes/$outwardcode");
    }

    /**
     * Get the address details from a random postcode
     *
     * @return object
     */
    public function getRandomPostcode()
    {
        return $this->getResponse("random/postcodes");
    }

    /**
     * Query the API for a given string
     *
     * @param string $query
     *
     * @return Collection
     *
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    public function query(string $query): Collection
    {
        $queryString = http_build_query(['q' => $query]);

        return collect($this->getResponse("postcodes?$queryString"));
    }

    /**
     * Get data for the postcodes nearest to the passed postcode
     *
     * @param string $postcode
     *
     * @return Collection
     *
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    public function nearest(string $postcode): Collection
    {
        return collect($this->getResponse("postcodes/$postcode/nearest"));
    }

    /**
     * Lookup a terminated postcode. Returns the postcode, year and month of termination.
     *
     * @param string $postcode
     *
     * @return object
     */
    public function getTerminatedPostcode($postcode)
    {
        return $this->getResponse("terminated_postcodes/$postcode");
    }

    /**
     * Autocomplete a postcode partial.
     *
     * @param string $partialPostcode
     *
     * @return Collection
     *
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    public function autocomplete(string $partialPostcode): Collection
    {
        return collect($this->getResponse("postcodes/$partialPostcode/autocomplete"));
    }

    /**
     * Get nearest outward codes for a given longitude & latitude
     *
     * @param float $longitude
     *
     * @param float $latitude
     * @return Collection
     *
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    public function nearestOutwardCodesForGivenLngAndLat(float $longitude, float $latitude): Collection
    {
        return collect($this->getResponse(sprintf(
            'outcodes?lon=%s&lat=%s',
            $longitude,
            $latitude
        )));
    }

    /**
     * Get information about nearest outcodes based on outward code
     *
     * @param string $outwardcode
     *
     * @param int $limit Needs to be less than 100
     * @param int $radius Needs to be less than 25,000m
     * @return Collection
     *
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    public function getNearestOutwardCode(
        string $outwardcode,
        int $limit = 10,
        int $radius = 5000
    ): Collection {
        $limit = ($limit > 100) ? 100 : $limit;
        $radius = ($radius > 100) ? 25000 : $radius;

        return collect($this->getResponse(
            "outcodes/$outwardcode/nearest?limit=$limit&radius=$radius"
        ));
    }

    /**
     * Get nearest postcodes for a given longitude & latitude
     *
     * @param float $longitude
     * @param float $latitude
     *
     * @return Collection
     *
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    public function nearestPostcodesForGivenLngAndLat(float $longitude, float $latitude): Collection
    {
        return collect($this->getResponse(sprintf(
            'postcodes?lon=%s&lat=%s',
            $longitude,
            $latitude
        )));
    }

    /**
     * Get the response and return the result object
     *
     * @param string|null $uri
     * @param string $method
     * @param array $data - data to be sent in post/patch/put request
     * @param array $options - array of options to be passed to curl, if $data is passed 'json' will be overwritten
     * @return mixed
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    protected function getResponse(
        string $uri = null,
        string $method = 'GET',
        array $data = [],
        array $options = []
    ) {
        $url = $this->url . $uri;

        if (!empty($data)) {
            $options['json'] = $data;
        }

        $request = $this->http->request(
            $method,
            $url,
            $options
        );

        return json_decode($request->getBody()->getContents())->result;
    }
}
